src/library/scala/collection/immutable/Stream.scala

```diff
@@ -1293,6 +1244,23 @@ object Stream extends SeqFactory[Stream] {
   private[immutable] def collectedTail[A, B, That](head: B, stream: Stream[A], pf: PartialFunction[A, B], bf: CanBuildFrom[Stream[A], B, That]) = {
     cons(head, stream.tail.collect(pf)(bf).asInstanceOf[Stream[B]])
   }
-}
-
 
+  /** An implementation of `FilterMonadic` allowing GC of the `Stream` as it is processed.
+    *
+    * Because this is not an inner class of `Stream`, it is now possible to offer explicit
+    * 'null after first use' semantics for the reference to the head of the original
+    * `Stream`, allowing GC to collect while the tail is still processing (see SI-8990).
+    */
```

Comment:
I had to read that a few times until I understood it. Worth it, though: Clever!

Just to confirm my understanding, you mean:

```scala
class StreamWithFilter[A](private var s: Stream[A], pred: A => Boolean) {
   lazy private val filtered = try { s filter pred } finally { s = null }
  def map[B, That](f: A => B)(implicit bf: CanBuildFrom[Stream[A], B, That]): That                                          = filtered.map(f)
  ..
}
```